pkg/sql/opt/memo/extract.go

```diff
@@ -540,3 +543,126 @@ func ExtractTailCalls(expr opt.Expr, tailCalls map[opt.ScalarExpr]struct{}) {
 		}
 	}
 }
+
+// ExtractExactScalarExprsForColumns extracts scalar expressions that tightly
+// constrain the given columns from the given filters. Unlike constraint
+// builder, this function:
+//   - constrains to scalar expressions instead of constants
+//   - always constrains tightly
+//   - in fact, always constrains _exactly_ (that is, to scalar expressions that
+//     produce not only equivalent values, but _exactly_ the same bytes)
+//
+// If a filter does not exactly constrain one of the given columns, or
+// references more than one of the given columns, the filter will be added to
+// remainingFilters. ExtractExactScalarExprsForColumns returns the mapping from
+// constrained columns to scalar expressions, the remaining filters, and a
+// boolean indicating whether all columns were constrained.
+func ExtractExactScalarExprsForColumns(
+	cols opt.ColSet,
+	filters FiltersExpr,
+	evalCtx *eval.Context,
+	md *opt.Metadata,
+	constructIsNotNull func(opt.ColumnID) FiltersItem,
+	constructFiltersItem func(opt.ScalarExpr) FiltersItem,
+) (constraints map[opt.ColumnID]opt.ScalarExpr, remainingFilters FiltersExpr, ok bool) {
+	constraints = make(map[opt.ColumnID]opt.ScalarExpr)
+
+	// useExprAsConstraint adds the scalar expression to the constraints map.
+	useExprAsConstraint := func(col opt.ColumnID, e opt.ScalarExpr) bool {
+		// Filters on other columns become remainingFilters.
+		if !cols.Contains(col) {
+			return false
+		}
+		// Only use the first constraining filter found. If there is already a
+		// constraining filter, any additional filters become part of
+		// remainingFilters.
+		if _, ok := constraints[col]; ok {
+			return false
+		}
+		// In the general case, the = operator (SQL equivalence) does not imply
+		// exactly the same bytes. The scalar expression is exactly the same when:
+		// - the column does not have composite type
+		// - the scalar type is identical to the column type OR the scalar is NULL
+		colType := md.ColumnMeta(col).Type
+		if colinfo.CanHaveCompositeKeyEncoding(colType) {
+			// TODO(michae2): If ExtractConstDatum(e) returns a non-composite datum,
+			// we could still constrain using this expression.
```

Comment:
Is this true? Could we have the value in the table be `1.000::DECIMAL` and the one in the filter `1::DECIMAL`?